internal/pkg/levee/propagation/propagation.go

```diff
@@ -248,6 +251,46 @@ func (prop *Propagation) visitOperands(n ssa.Node, maxInstrReached map[*ssa.Basi
 	}
 }
 
+func (prop *Propagation) visitSelect(sel *ssa.Select, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
+	// The Send operations have to be processed before the Recv operations in case a channel
+	// is written to in one branch and read from in a different branch. This matters if the
+	// branches are reachable from each other, e.g. when the select is in a loop.
+
+	for _, s := range sel.States {
+		if s.Dir == types.SendOnly {
+			// If the sent value (Send) is tainted, propagate taint to the channel
+			if prop.marked[s.Send.(ssa.Node)] {
+				prop.dfs(s.Chan.(ssa.Node), maxInstrReached, lastBlockVisited, false)
+			}
+		}
+	}
+
+	if sel.Referrers() == nil {
+		return
+	}
+
+	recvStateCount := 0
+	for _, s := range sel.States {
+		if s.Dir == types.RecvOnly {
+			recvStateCount++
+			if !prop.marked[s.Chan.(ssa.Node)] {
+				continue
+			}
+			// If the channel is tainted, propagate taint to the appropriate Extract
+			for _, r := range *sel.Referrers() {
+				e, ok := r.(*ssa.Extract)
+				// Select returns a tuple whose first 2 elements are irrelevant for our
+				// analysis. The other elements map 1:1 with each of the recv states.
+				// See the docs for ssa.Extract for more details.
```

Comment:
Something felt a little off-kilter about this docstring explaining the `recvStateCount` that is used in the wrapping blocks.  And as mentioned elsewhere, the separation of send and recv felt a bit strange to me...

I was poking at it as I was trying to understand this PR, and I wound up with this implementation of `visitSelect`.  What do you think of the readability of extracting this counter as a bit of preprocessing, and combing the actual assessment of each `case` within a single loop?

```golang
func (prop *Propagation) visitSelect(sel *ssa.Select, maxInstrReached map[*ssa.BasicBlock]int, lastBlockVisited *ssa.BasicBlock) {
	// Comment about extracted tuple and why count starts at 2.  Reference docs.
	// This could also be extracted as a separate method if that would further improve readability
	var recvInds = map[int]int{}
	count := 2
	for i, s := range sel.States {
		if s.Dir == types.RecvOnly {
			recvInds[i] = count
			count++
		}
	}

	for i, s := range sel.States {
		switch {
		// High-level comment about tainting chans
		case s.Dir == types.SendOnly && prop.marked[s.Send.(ssa.Node)]:
			prop.dfs(s.Chan.(ssa.Node), maxInstrReached, lastBlockVisited, false)

		// High-level comment about recieving tainted value via chans
		case s.Dir == types.RecvOnly && prop.marked[s.Chan.(ssa.Node)]:
			if sel.Referrers() == nil {
				continue
			}
			// Implementation comment about extraction referrer
			for _, r := range *sel.Referrers() {
				e, ok := r.(*ssa.Extract)
				if !ok || e.Index != recvInds[i] {
					continue
				}
				prop.dfs(e, maxInstrReached, lastBlockVisited, false)
			}
		}
	}
}
```

It puts the nil-safety check nearer its usage, too, so even if it ends up being unreachable in practice, your argument for keeping it there as safety is more apparent in the implementation.

Thoughts?

Reply:
I do like having the two cases in the same loop + switch. My perception that the sends have to be visited before the recvs may have been caused by the earlier `count++` bug that is now fixed.

I also like the idea of computing the indices in advance. I took a slightly different approach w.r.t. to the implementation details. Starting the count at 2 doesn't make sense to me.